### PR TITLE
Propagate prereleases to newer majors

### LINKED ISSUE #126: Pre-releases are not propagated to newer majors
I want to do a pre-release of makd v1.11.0, but when doing so, there is no pre-release of v2.1.0, which is needed because some users that are willing to test the RC are already using the v2.x.x releases and can't downgrade.

This is what I get from the current MakD:
```
makd:v1.x.x $%= ✔ $ neptune-release -p
Please enter the upstream remote name.
Will default to [origin]: 
We are on branch v1.x.x
Detected release v1.11.0-rc1
Warning: No corresponding milestone found for v1.11.0-rc1
Actions to be done:
 * Creating annotated tag v1.11.0-rc1 (["git", "tag", "-F-", "v1.11.0-rc1", "v1.x.x"])
 * Checkout v2.x.x locally (["git", "checkout", "v2.x.x"])
 * Merge v1.11.0-rc1 into v2.x.x (["git", "merge", "--no-ff", "-m", "Merge tag v1.11.0-rc1 into v2.x.x", "v1.11.0-rc1"])
 * Checkout previous branch v1.x.x (["git", "checkouts", "v1.x.x"])
 * Checkout original branch v1.x.x (["git", "checkout", "v1.x.x"])

Would you like to continue? y/n: n
```

### PR DESCRIPTION
**Pull request by mathias-baumann-sociomantic**
fix #126

Not really sure about this, but luca requested it. I didn't think we'd want this
behavior tbh.

### CODE REVIEW

**Comment by leandro-lucarella-sociomantic** [comment]
BTW, add a test if it's not too complicated?

**Comment by mathias-baumann-sociomantic** [comment]
Yeah, we agreed it goes in. It might need some more work though

**Comment by leandro-lucarella-sociomantic** [comment]
Does it make sense to try to test it as it is with makd? I need to make a release. Otherwise I will skip the pre-release for it this time.

**Comment by mathias-baumann-sociomantic** [comment]
well, the worst thing that can happen is that it tries to create a non-rc version in the newer majors

**Comment by leandro-lucarella-sociomantic** [comment]
Yup, that's what happens:
```
makd:v1.x.x $%= 130 $ neptune-release -p 
Please enter the upstream remote name.
Will default to [origin]: 
We are on branch v1.x.x
Detected release v1.11.0-rc1
Warning: No corresponding milestone found for v1.11.0-rc1
Actions to be done:
 * Creating annotated tag v1.11.0-rc1 (["git", "tag", "-F-", "v1.11.0-rc1", "v1.x.x"])
 * Checkout v2.x.x locally (["git", "checkout", "v2.x.x"])
 * Merge v1.11.0-rc1 into v2.x.x (["git", "merge", "--no-ff", "-m", "Merge tag v1.11.0-rc1 into v2.x.x", "v1.11.0-rc1"])
 * Checkout previous branch v1.x.x (["git", "checkouts", "v1.x.x"])
 * Checkout next branch v2.x.x (["git", "checkout", "v2.x.x"])
 * Creating annotated tag v2.1.0 (["git", "tag", "-F-", "v2.1.0", "v2.x.x"])
 * Create tracking branch v2.1.x (["git", "branch", "v2.1.x", "v2.1.0"])
 * Checkout original branch v1.x.x (["git", "checkout", "v1.x.x"])

Would you like to continue? y/n: 
```

**Comment by leandro-lucarella-sociomantic** [comment]
I will go with normal releases this time.

**Comment by leandro-lucarella-sociomantic** [comment]
BTW, without this IMHO the pre-release feature is useless, so increasing the priority...

**Comment by mathias-baumann-sociomantic** [comment]
Pushed a fix. Now based on https://github.com/sociomantic-tsunami/neptune/pull/127

File: src/semver/Version.d [hunk 1]
```diff
@@ -154,14 +178,20 @@ struct Version
         auto vers = [
             Version.parse("v2.2.0"),
             Version.parse("v1.3.0"),
+            Version.parse("v1.2.4-rc.4"),
             Version.parse("v1.2.3"),
             Version.parse("v1.2.4"),
+            Version.parse("v1.2.4-alpha1"),
```
Comment (leandro-lucarella-sociomantic) [inline src/semver/Version.d]: The dot notation should be also done for alphas (`alpha.1`, although if this is testing that the tool doesn't break with just any arbitrary prerelease string, then good).

File: src/release/versionHelper.d [hunk 2]
```diff
@@ -320,40 +320,59 @@ public bool needMinorRelease ( A, B ) ( A matching_major, B matching_minor,
...
-    return false;
+public Version getNextMinor ( Version last )
+{
+    import release.options;
+    import std.algorithm;
+
+    int rc = 1;
+
+    import std.stdio;
+    writefln("LAST VERSION: %s", last);
```
Comment (leandro-lucarella-sociomantic) [inline src/release/versionHelper.d]: I guess this slipped in and should be removed?

File: src/release/main.d [hunk 3]
```diff
@@ -639,7 +639,7 @@ ActionList prepareMinorRelease ( ref HTTPConnection con, ref Repository repo,
 
                 // Checkout previous branch so we can remove the rel notes
                 list.actions ~= new LocalAction(
-                                 ["git", "checkouts", current_branch.toString],
+                                 ["git", "checkout", current_branch.toString],
```
Comment (leandro-lucarella-sociomantic) [inline src/release/main.d]: That was probably me... whoops!

How hard would it be to add a test for this so I don't break it again? :grimacing:

**Comment by leandro-lucarella-sociomantic** [comment]
> Pushed a fix. Now based on #127

Oh, I reviewed all commits here.

File: src/semver/Version.d (hunk 1 above)
Comment (mathias-baumann-sociomantic) [reply src/semver/Version.d]: Well, in this case the intention was to test for arbitrary strings. There is no logic for doing that processing for alpha yet

File: src/release/versionHelper.d (hunk 2 above)
Comment (mathias-baumann-sociomantic) [reply src/release/versionHelper.d]: ah, yes

File: src/release/versionHelper.d [hunk 4]
```diff
@@ -320,40 +320,59 @@ public bool needMinorRelease ( A, B ) ( A matching_major, B matching_minor,
...
-                new_version.minor = minor + 1;
+/*******************************************************************************
 
-            if (options.pre_release)
-            {
-                import std.format;
-                new_version.prerelease = format("%s%s", RCPrefix, rc);
-            }
+    Find out the minor version that would follow 'last'.
+    Considers RC releases when the accordin global option was given.
```
Comment (mathias-baumann-sociomantic) [inline src/release/versionHelper.d]: typo, `according` misses g

File: src/release/main.d (hunk 3 above)
Comment (mathias-baumann-sociomantic) [reply src/release/main.d]: This is actually covered by the tests added here

**Comment by mathias-baumann-sociomantic** [comment]
Updated:
* fixed typo
* removed debug print

File: integrationtest/common/TestCase.d [hunk 5]
```diff
@@ -226,24 +226,27 @@ class TestCase
...
-            auto file = File(format("%s/relnotes.md", this.common_data), "r");
+            auto file = File(path, "r");
             auto fsize = file.size();
             assert(fsize < 1024 * 16, "relnotes file unexpectedly large!");
 
             correct_relnotes = strip(file.rawRead(new char[fsize]));
 
             auto gh_rel = this.fake_github.releases.find!(a=>a.name == ver);
-            assert(!gh_rel.empty, "Release not found on gh fake server!");
+            assert(!gh_rel.empty, "Release "~ ver ~" not found on gh fake server!");
```
Comment (leandro-lucarella-sociomantic) [inline integrationtest/common/TestCase.d]: Not a biggie but this change looks unrelated to this commit.

**Comment by mathias-baumann-sociomantic** [comment]
ping pong?

**Comment by leandro-lucarella-sociomantic** [comment]
I just released tangort v1.8.0-rc.1 and all seems to have gone according to plan!

File: integrationtest/common/TestCase.d (hunk 5 above)
Comment (leandro-lucarella-sociomantic) [reply integrationtest/common/TestCase.d]: ping

**Comment by mathias-baumann-sociomantic** [comment]
I will need to know what message that would be

**Comment by leandro-lucarella-sociomantic** [comment]
I haven't tested it with a project that has to propagate though.

**Comment by leandro-lucarella-sociomantic** [comment]
But for me it's OK to merge if it looks good, if it has bugs they can be handled when they are discovered.
